pjsip/src/pjsip/sip_transport.c

```diff
@@ -2066,6 +2067,26 @@ PJ_DEF(pj_ssize_t) pjsip_tpmgr_receive_packet( pjsip_tpmgr *mgr,
         /* For TCP transport, check if the whole message has been received. */
         if ((tr->flag & PJSIP_TRANSPORT_DATAGRAM) == 0) {
             pj_status_t msg_status;
+
+            if (mgr->tp_rx_data_cb) {
+                pjsip_tp_rx_data rd;
+                pj_bzero(&rd, sizeof(rd));
+                rd.tp = tr;
+                rd.data = current_pkt;
+                rd.len = remaining_len;
+                rd.status = PJ_EIGNORED;
+
+                if ((*mgr->tp_rx_data_cb)(&rd) == PJ_SUCCESS) {
+                    /* Sanity check */
+                    if (rd.len > remaining_len)
+                        rd.len = remaining_len;
+                    total_processed += rd.len;
+                    current_pkt += rd.len;
+                    remaining_len -= rd.len;
+                    continue;
```

Comment:
IMHO, perhaps custom processings here is a bit too much, some default nice features such as initial idle timer, parsing error logging, drop data callback, will be gone (need to be re-implemented by app if wanted). Moreover, if app is allowed to do custom rdata processings by itself, I think we need to describe more things in the docs, e.g:
- app may need to invoke `pjsip_endpt_process_rx_data()` since `mgr->on_rx_msg/endpt_on_rx_msg()` is no longer called.
- describe all processings done by the lib, so the developer will be aware, what they do or are for, which processings are mandatory/optional/modifiable.

Perhaps allowing app to verify the data or shift the start of the data (e.g: for skipping garbage) is sufficient for now?

Reply:
Could you clarify:
```
      remaining_len = rd.len;

      msg_fragment_size = remaining_len - rd.len;
      total_processed += msg_fragment_size;
      current_pkt += msg_fragment_size;
```
this means `msg_fragment_size` is 0, and `current_pkt` will not be increased.

Reply:
> this means `msg_fragment_size` is 0, and `current_pkt` will not be increased.

Good catch. I incorrectly moved it to the top of the block.